policy: Add validation and docs for TLS SNI ServerNames

sayboras opened the pull request:
### Description 

After the below PR, TLS SNI server names are now supporting prefix wildcard match, this PR is to update the docs and basic validation for the same.

Relates: https://github.com/cilium/proxy/pull/1242

Note: Wait for https://github.com/cilium/cilium/pull/38603 to land in first

---
### Testing

Testing was done locally as per below

```
$ test_name=client-egress-tls-sni
$ cilium connectivity test --test=$test_name
[=] [cilium-test-1] Test [client-egress-tls-sni] [75/119]
.........
[=] [cilium-test-1] Test [client-egress-tls-sni-denied] [76/119]
.........
[=] [cilium-test-1] Test [client-egress-tls-sni-wildcard] [77/119]
.........
[=] [cilium-test-1] Test [client-egress-tls-sni-wildcard-denied] [78/119]
...
[=] [cilium-test-1] Test [client-egress-tls-sni-double-wildcard] [79/119]
.........
[=] [cilium-test-1] Test [client-egress-tls-sni-double-wildcard-denied] [80/119]
...
✅ [cilium-test-1] All 6 tests (42 actions) successful, 113 tests skipped, 0 scenarios skipped.
```

sayboras committed:
api: Add validation for SNI ServerNames

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
connectivity: Add test for TLS wildcard SNI

Signed-off-by: Tam Mach <tam.mach@cilium.io>

jrajahalme commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -986,8 +986,25 @@ spec:
+                                  - `*cilium.io` matches cilium.io and all subdomains ends with "cilium.io" except those containing "." separator, subcilium.io and sub-cilium.io match,
+                                    www.cilium.io and blog.cilium.io does not.
+                              maxLength: 255
+                              pattern: ^([-a-zA-Z0-9_*]+[.]?)+$
```
It should be possible to validate that the '*' characters, if any, are at the beginning, i.e.:
```
pattern: ^(\*?\*\.)?([-a-zA-Z0-9_]+\.?)+$
```

jrajahalme commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -986,8 +986,25 @@ spec:
+                                ServerName allows using prefix only wildcards to match DNS names.
+
+                                - "*" matches 0 or more DNS valid characters, and may occur anywhere in
+                                the pattern. As a special case a "*" as the leftmost character, without a
```
Should this be "and may only occur at the beginning of the pattern" instead?

jrajahalme commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -986,8 +986,25 @@ spec:
+                                Examples:
+                                  - `*.cilium.io` matches exactly one subdomain of cilium at that level www.cilium.io and blog.cilium.io match, cilium.io and google.com do not.
+                                  - `**.cilium.io` matches more than one subdomain of cilium, e.g. sub1.sub2.cilium.io and sub.cilium.io match, cilium.io do not.
+                                  - `*cilium.io` matches cilium.io and all subdomains ends with "cilium.io" except those containing "." separator, subcilium.io and sub-cilium.io match,
```
This one may actually be dangerous. What if the name happens to be "evil-cilium.io"? There is no (ownership) relation between it and "cilium.io", so IMO this kind of prefix matching should not be allowed.

sayboras replied:
Thanks, I have removed this in related cilium/proxy PR.

jrajahalme commented on pkg/policy/api/l4.go:
```diff
@@ -169,6 +169,24 @@ type Listener struct {
+// ServerName allows using prefix only wildcards to match DNS names.
+//
+// - "*" matches 0 or more DNS valid characters, and may occur anywhere in
+// the pattern. As a special case a "*" as the leftmost character, without a
```
also here

jrajahalme commented on pkg/policy/api/l4.go:
```diff
@@ -169,6 +169,24 @@ type Listener struct {
+// Examples:
+//   - `*.cilium.io` matches exactly one subdomain of cilium at that level www.cilium.io and blog.cilium.io match, cilium.io and google.com do not.
+//   - `**.cilium.io` matches more than one subdomain of cilium, e.g. sub1.sub2.cilium.io and sub.cilium.io match, cilium.io do not.
+//   - `*cilium.io` matches cilium.io and all subdomains ends with "cilium.io" except those containing "." separator, subcilium.io and sub-cilium.io match,
```
and here

jrajahalme commented on pkg/policy/api/l4.go:
```diff
@@ -169,6 +169,24 @@ type Listener struct {
+//     www.cilium.io and blog.cilium.io does not.
+//
+// +kubebuilder:validation:MaxLength=255
+// +kubebuilder:validation:Pattern=`^([-a-zA-Z0-9_*]+[.]?)+$`
```
and here :-)

jrajahalme reviewed (changes requested):
Some early comments below

sayboras committed:
check: Skip warn log for TLS context

Similarly to external target, this is to skip the warning log for
external other target.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

derailed commented on cilium-cli/connectivity/builder/client_egress_tls_sni.go:
```diff
@@ -49,6 +49,31 @@ func clientEgressTlsSniTest(ct *check.ConnectivityTest, templates map[string]str
 			return check.ResultDefaultDenyEgressDrop, check.ResultNone
 		})
+
+	yamlFile = templates["clientEgressTLSSNIWildcardPolicyYAML"]
```
nit: Perhaps we could add a check on the map access in case the key is not there?

sayboras replied:
:thinking: probably we can take it separately, as all the template map usage didn't have existence check :cry:

derailed reviewed (approved):
@sayboras Nice work Tam!

jrajahalme reviewed (commented):
Should we support testing with `**.` prefix instead of `*.*.` prefix?
